### PR TITLE
LPS-121375 The Anonymous User Cannot be Changed

### PR DESCRIPTION
**Pull request by erickmonteiroo**

### CODE REVIEW

**Comment by erickmonteiroo** [comment]
ci:test:sf

File: modules/apps/user-associated-data/user-associated-data-web/src/main/java/com/liferay/user/associated/data/web/internal/upgrade/v1_1_0/ConfigurationUpgradeProcess.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+	private void _updateDictionaryToScoped(
+			PreparedStatement preparedStatement, ResultSet resultSet)
+		throws SQLException {
+
+		String oldDictionary = resultSet.getString("dictionary");
+
+		String newDictionary = StringUtil.replace(
+			oldDictionary, "AnonymousUserConfiguration",
+			"AnonymousUserConfiguration.scoped");
```
Comment (pei-jung) [inline modules/apps/user-associated-data/user-associated-data-web/src/main/java/com/liferay/user/associated/data/web/internal/upgrade/v1_1_0/ConfigurationUpgradeProcess.java:93]: Same here. Use fully qualified class name

File: modules/apps/user-associated-data/user-associated-data-web/src/main/java/com/liferay/user/associated/data/web/internal/upgrade/v1_1_0/ConfigurationUpgradeProcess.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+		preparedStatement.setString(1, newDictionary);
```
Comment (pei-jung) [inline modules/apps/user-associated-data/user-associated-data-web/src/main/java/com/liferay/user/associated/data/web/internal/upgrade/v1_1_0/ConfigurationUpgradeProcess.java:95]: Inline

**Review by pei-jung — CHANGES_REQUESTED** [review]
@erickmonteiroo nice work on the upgrade! A couple minor SF that needs to be fixed. Otherwise, everything looks good. If possible, can we also include a test case for this?

**Comment by drewbrokke** [comment]
@erickmonteiroo you're going to need to add a ticket number to that latest commit, otherwise CI will complain

**Comment by erickmonteiroo** [comment]
> @erickmonteiroo you're going to need to add a ticket number to that latest commit, otherwise CI will complain

@drewbrokke Not intentional, I wasn't supposed to do the push right now. Sorry about that.

**Comment by erickmonteiroo** [comment]
ci:test:sf

**Comment by erickmonteiroo** [comment]
ci:test:relevant

**Comment by erickmonteiroo** [comment]
ci:test:sf

**Comment by erickmonteiroo** [comment]
ci:test:relevant

**Comment by erickmonteiroo** [comment]
ci:test:upgrade

**Comment by Sharryxue** [comment]
Related failure found in enterprise db. Hello @erickmonteiroo, would you please take a look into the issues, thanks.
- Oracle, refer [test](https://testray.liferay.com/home/-/testray/case_results/1546912779), [console](https://testray.liferay.com/reports/production/logs/2021-07/test-1-23/test-portal-acceptance-pullrequest(master)/7931/functional-upgrade-tomcat90-oracle193-jdk8/4/3/jenkins-console.txt.gz)
```
[java] 2021-07-13 22:25:12.287 INFO  [main][UpgradeProcess:126] Failed upgrade process com.liferay.user.associated.data.web.internal.upgrade.v1_1_0.ConfigurationUpgradeProcess in 6 ms
[java] 2021-07-13 22:25:12.293 ERROR [main][UpgradeStepRegistratorTracker:143] Failed upgrade process for module com.liferay.user.associated.data.web
[java] com.liferay.portal.kernel.upgrade.UpgradeException: java.sql.SQLSyntaxErrorException: ORA-00904: "com.liferay.user.associated.data.web.internal.configuration.AnonymousUserConfiguration.scoped%": invalid identifier
[java] 
[java] 	at com.liferay.portal.kernel.upgrade.UpgradeProcess.upgrade(UpgradeProcess.java:120) ~[portal-kernel.jar:?]
[java] 	at com.liferay.portal.kernel.upgrade.UpgradeProcess.upgrade(UpgradeProcess.java:138) ~[portal-kernel.jar:?]
[java] 	at com.liferay.portal.upgrade.internal.executor.UpgradeExecutor$UpgradeInfosRunnable.run(UpgradeExecutor.java:199) ~[bundleFile:?]
```
- PostgreSQL, refer [tes](https://testray.liferay.com/home/-/testray/case_results/1546913184)t, [console](https://testray.liferay.com/reports/production/logs/2021-07/test-1-23/test-portal-acceptance-pullrequest(master)/7931/functional-upgrade-tomcat90-postgresql11-jdk8/5/13/jenkins-console.txt.gz)
```
 [parallel] 2021-07-14 02:39:43.433 INFO  [main][UpgradeProcess:126] Failed upgrade process com.liferay.user.associated.data.web.internal.upgrade.v1_1_0.ConfigurationUpgradeProcess in 4 ms
 [parallel] 2021-07-14 02:39:43.436 ERROR [main][UpgradeStepRegistratorTracker:143] Failed upgrade process for module com.liferay.user.associated.data.web
 [parallel] com.liferay.portal.kernel.upgrade.UpgradeException: org.postgresql.util.PSQLException: ERROR: column "com.liferay.user.associated.data.web.internal.configuration.Ano" does not exist
 [parallel]   Position: 57
 [parallel] 	at com.liferay.portal.kernel.upgrade.UpgradeProcess.upgrade(UpgradeProcess.java:120) ~[portal-kernel.jar:?]
 [parallel] 	at com.liferay.portal.kernel.upgrade.UpgradeProcess.upgrade(UpgradeProcess.java:138) ~[portal-kernel.jar:?]
```

 - db2, refer [test](https://testray.liferay.com/zh/home/-/testray/case_results/1546911243), [console](https://testray.liferay.com/reports/production/logs/2021-07/test-1-23/test-portal-acceptance-pullrequest(master)/7931/functional-upgrade-tomcat90-db2111-jdk8/2/5/jenkins-console.txt.gz)
```
[java] 2021-07-13 21:36:43.106 INFO  [main][UpgradeProcess:111] Upgrading com.liferay.user.associated.data.web.internal.upgrade.v1_1_0.ConfigurationUpgradeProcess
[java] 2021-07-13 21:36:43.108 INFO  [main][UpgradeProcess:126] Failed upgrade process com.liferay.user.associated.data.web.internal.upgrade.v1_1_0.ConfigurationUpgradeProcess in 2 ms
[java] 2021-07-13 21:36:43.112 ERROR [main][UpgradeStepRegistratorTracker:143] Failed upgrade process for module com.liferay.user.associated.data.web
[java] com.liferay.portal.kernel.upgrade.UpgradeException: com.ibm.db2.jcc.am.SqlSyntaxErrorException: DB2 SQL Error: SQLCODE=-206, SQLSTATE=42703, SQLERRMC=com.liferay.user.associated.data.web.internal.configuration.AnonymousU, DRIVER=4.24.92
[java] 	at com.liferay.portal.kernel.upgrade.UpgradeProcess.upgrade(UpgradeProcess.java:120) ~[portal-kernel.jar:?]
[java] 	at com.liferay.portal.kernel.upgrade.UpgradeProcess.upgrade(UpgradeProcess.java:138) ~[portal-kernel.jar:?]
```
- sqlserver2017, refer [test](https://testray.liferay.com/home/-/testray/case_results/1546913517), [console](https://testray.liferay.com/reports/production/logs/2021-07/test-1-23/test-portal-acceptance-pullrequest(master)/7931/functional-upgrade-tomcat90-sqlserver2017-jdk8/1/12/jenkins-console.txt.gz)
```
 [parallel] 2021-07-13 20:21:49.515 INFO  [main][UpgradeProcess:126] Failed upgrade process com.liferay.user.associated.data.web.internal.upgrade.v1_1_0.ConfigurationUpgradeProcess in 2 ms
 [parallel] 2021-07-13 20:21:49.522 ERROR [main][UpgradeStepRegistratorTracker:143] Failed upgrade process for module com.liferay.user.associated.data.web
 [parallel] com.liferay.portal.kernel.upgrade.UpgradeException: com.microsoft.sqlserver.jdbc.SQLServerException: Invalid column name 'com.liferay.user.associated.data.web.internal.configuration.AnonymousUserConfiguration%'.
 [parallel] 	at com.liferay.portal.kernel.upgrade.UpgradeProcess.upgrade(UpgradeProcess.java:120) ~[portal-kernel.jar:?]
 [parallel] 	at com.liferay.portal.kernel.upgrade.UpgradeProcess.upgrade(UpgradeProcess.java:138) ~[portal-kernel.jar:?]
 [parallel] 	at com.liferay.portal.upgrade.internal.executor.UpgradeExecutor$UpgradeInfosRunnable.run(UpgradeExecutor.java:199) ~[bundleFile:?]
```

**Comment by erickmonteiroo** [comment]
ci:test:upgrade

**Comment by erickmonteiroo** [comment]
ci:test:relevant

**Comment by erickmonteiroo** [comment]
ci:test:relevant

**Comment by erickmonteiroo** [comment]
ci:test:upgrade

**Comment by Sharryxue** [comment]
- Find following error on all environments, refer [console](https://testray.liferay.com/reports/production/logs/2021-07/test-1-28/test-portal-acceptance-pullrequest(master)/7/functional-upgrade-tomcat90-db2111-jdk8/1/20/jenkins-console.txt.gz), would you please double check @erickmonteiroo, thanks 
```
     [java] 2021-07-19 21:29:02.025 INFO  [main][UpgradeProcess:126] Failed upgrade process com.liferay.user.associated.data.web.internal.upgrade.v1_1_0.ConfigurationUpgradeProcess in 5 ms
     [java] 2021-07-19 21:29:02.029 ERROR [main][UpgradeStepRegistratorTracker:143] Failed upgrade process for module com.liferay.user.associated.data.web
     [java] com.liferay.portal.kernel.upgrade.UpgradeException: com.ibm.db2.jcc.am.SqlSyntaxErrorException: DB2 SQL Error: SQLCODE=-440, SQLSTATE=42884, SQLERRMC=CAST_CLOB_TEXT;FUNCTION, DRIVER=4.24.92
     [java] 	at com.liferay.portal.kernel.upgrade.UpgradeProcess.upgrade(UpgradeProcess.java:120) ~[portal-kernel.jar:?]
     [java] 	at com.liferay.portal.kernel.upgrade.UpgradeProcess.upgrade(UpgradeProcess.java:138) ~[portal-kernel.jar:?]
```

**Comment by erickmonteiroo** [comment]
ci:test:sf

**Comment by erickmonteiroo** [comment]
ci:test:relevant

**Comment by erickmonteiroo** [comment]
ci:test:upgrade

**Comment by patriciajeanperez** [comment]
Upgrade failures are not related. The [previous failures](https://github.com/liferay-usm/liferay-portal/pull/123#issuecomment-879862755) on user associated data are no longer present. @erickmonteiroo

File: modules/apps/user-associated-data/user-associated-data-test/src/testIntegration/java/com/liferay/user/associated/data/anonymizer/test/UADAnonymousUserProviderTest.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+public class UADAnonymousUserProviderTest {
+
+	@ClassRule
+	@Rule
+	public static final AggregateTestRule aggregateTestRule =
+		new LiferayIntegrationTestRule();
+
+	@Test
+	public void testGetAnonymousUser() throws Exception {
+		User anonymousUser1 = UserTestUtil.addUser();
```
Comment (pei-jung) [inline modules/apps/user-associated-data/user-associated-data-test/src/testIntegration/java/com/liferay/user/associated/data/anonymizer/test/UADAnonymousUserProviderTest.java:44]: Why is this called anonymous user? It's simply adding a regular user here.

File: modules/apps/user-associated-data/user-associated-data-test/src/testIntegration/java/com/liferay/user/associated/data/anonymizer/test/UADAnonymousUserProviderTest.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+		Assert.assertTrue(
+			_uadAnonymousUserProvider.isAnonymousUser(
+				_uadAnonymousUserProvider.getAnonymousUser(
+					anonymousUser1.getCompanyId())));
+
+		User anonymousUser2 = UserTestUtil.addUser();
```
Comment (pei-jung) [inline modules/apps/user-associated-data/user-associated-data-test/src/testIntegration/java/com/liferay/user/associated/data/anonymizer/test/UADAnonymousUserProviderTest.java:51]: Why do we need to create a second user here?

File: modules/apps/user-associated-data/user-associated-data-test/src/testIntegration/java/com/liferay/user/associated/data/anonymizer/test/UADAnonymousUserProviderTest.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+		Assert.assertTrue(
+			_uadAnonymousUserProvider.isAnonymousUser(
+				_uadAnonymousUserProvider.getAnonymousUser(
+					anonymousUser2.getCompanyId())));
```
Comment (pei-jung) [inline modules/apps/user-associated-data/user-associated-data-test/src/testIntegration/java/com/liferay/user/associated/data/anonymizer/test/UADAnonymousUserProviderTest.java:56]: What are we trying to verify here? Isn't this a same check as line 46-49?

Both anonymousUser1 and anonymousUser2 are in the same company. You are just calling `_uadAnonymousUserProvider.getAnonymousUser` twice.

File: modules/apps/user-associated-data/user-associated-data-test/src/testIntegration/java/com/liferay/user/associated/data/anonymizer/test/UADAnonymousUserProviderTest.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+		Assert.assertFalse(
+			_uadAnonymousUserProvider.isAnonymousUser(anonymousUser1));
+	}
```
Comment (pei-jung) [inline modules/apps/user-associated-data/user-associated-data-test/src/testIntegration/java/com/liferay/user/associated/data/anonymizer/test/UADAnonymousUserProviderTest.java:60]: I don't quite understand what this test is trying to do. Tests should either replicate the steps to reproduce in the ticket or recreate the issue that you're trying to fix.

**Review by pei-jung — CHANGES_REQUESTED** [review]
@erickmonteiroo Great job with this change! Overall everything looks good. The test needs to be more comprehensive and clear of what it does. Please see my comments and let me know if you have any questions. Thanks!

**Comment by erickmonteiroo** [comment]
ci:test:sf

**Comment by erickmonteiroo** [comment]
ci:test:relevant

**Comment by erickmonteiroo** [comment]
ci:test:upgrade

**Comment by erickmonteiroo** [comment]
ci:stop

**Comment by erickmonteiroo** [comment]
ci:test:sf

**Comment by erickmonteiroo** [comment]
ci:test:sf

**Comment by erickmonteiroo** [comment]
ci:test:relevant

**Comment by erickmonteiroo** [comment]
ci:test:relevant

**Comment by erickmonteiroo** [comment]
ci:test:upgrade

**Comment by Sharryxue** [comment]
No related failure found in both ci:test:relevant and ci:test:upgrade suite.
The [semantic-versioning-jdk8/0](https://test-1-28.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/1322/) issue not related and fixed in [commit](https://github.com/liferay/liferay-portal/commit/a88f2082183e2dd8b021a07efcfb02b0f4910e6e).

File: modules/apps/user-associated-data/user-associated-data-web/src/main/java/com/liferay/user/associated/data/web/internal/upgrade/v1_1_0/ConfigurationUpgradeProcess.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+import com.liferay.user.associated.data.configuration.AnonymousUserConfiguration;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+/**
+ * @author Erick Monteiro
+ */
+public class ConfigurationUpgradeProcess extends UpgradeProcess {
```
Comment (drewbrokke) [inline modules/apps/user-associated-data/user-associated-data-web/src/main/java/com/liferay/user/associated/data/web/internal/upgrade/v1_1_0/ConfigurationUpgradeProcess.java:30]: We should prefer using `ConfigurationAdmin` for upgrade processes relating to the `Configuration_` table.

**Review by drewbrokke — CHANGES_REQUESTED** [review]
@erickmonteiroo This is looking good but we still need to refine it a little. Please see my inline comments.

**Comment by drewbrokke** [comment]
@erickmonteiroo will you also please remove the `companyId` method from the `AnonymousUserConfiguration`? It is not needed now that we are instance scoping the config and `companyId` will be injected automatically.

**Comment by erickmonteiroo** [comment]
@drewbrokke I'll be closing this and sending a new one.

File: modules/apps/user-associated-data/user-associated-data-api/src/main/java/com/liferay/user/associated/data/configuration/AnonymousUserConfigurationRetriever.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+	 * Get a Configuration Optional for the given company
+	 * @param  companyId the company ID
+	 * @return the anonymous user
+	 * @throws IOException if an exception occurred
+	 * @throws InvalidSyntaxException if an exception occurred
+	 */
+	public Optional<Configuration> getOptional(long companyId, long userId)
+		throws InvalidSyntaxException, IOException;
+
+}
```
[the remark this answers is not in the document]
Comment (erickmonteiroo) [reply modules/apps/user-associated-data/user-associated-data-api/src/main/java/com/liferay/user/associated/data/configuration/AnonymousUserConfigurationRetriever.java:50]: This one need to be in the API in order to perform the integration tests.

[the remark this answers is not in the document]
Comment (drewbrokke) [reply modules/apps/user-associated-data/user-associated-data-api/src/main/java/com/liferay/user/associated/data/configuration/AnonymousUserConfigurationRetriever.java:50]: Can we instead do a manual query on ConfigAdmin?

**Pull request closed without merging**